csharp/ql/src/semmle/code/csharp/PrintAst.qll

```diff
@@ -0,0 +1,606 @@
+    elem.fromSource() and
+    this.selectedFile(getRepresentativeLocation(elem).getFile())
+  }
+}
+
+private predicate shouldPrint(Element elem) {
+  exists(PrintAstConfiguration config | config.shouldPrint(elem))
+}
+
+private predicate selectedFile(File file) {
+  exists(PrintAstConfiguration config | config.selectedFile(file))
+}
+
+private int attributableOffset(Attributable attributable) {
+  if exists(attributable.getAnAttribute()) then result = 1 else result = 0
+}
+
+private int parameterizableTypeOffset(ValueOrRefType type) {
+  if type instanceof Parameterizable and type.(Parameterizable).getNumberOfParameters() > 0
+  then result = 1
+  else result = 0
+}
+
+private int parameterizableDeclarationOffset(DeclarationWithAccessors declaration) {
+  if declaration instanceof Indexer and declaration.(Parameterizable).getNumberOfParameters() > 0
+  then result = 1
+  else result = 0
+}
+
+private int unboundGenericOffset(ValueOrRefType type) {
+  if type instanceof UnboundGeneric then result = 1 else result = 0
+}
+
+private int assignableOffset(AssignableMember assignable) {
+  if assignable.(Property).hasInitializer() then result = 1 else result = 0
+}
+
+private predicate isInsideUnneededAttributable(Attributable attributable) {
+  isInsideUnneededType(attributable.(Field).getDeclaringType()) or
+  isInsideUnneededParameterizable(attributable.(Parameter).getDeclaringElement()) or
+  isInsideUnneededCallable(attributable.(Callable)) or
+  isInsideUnneededType(attributable.(DeclarationWithAccessors).getDeclaringType()) or
+  isInsideUnneededType(attributable.(ValueOrRefType))
+}
+
+private predicate isInsideUnneededUnboundGeneric(UnboundGeneric unboundGeneric) {
+  isInsideUnneededAttributable(unboundGeneric) or
+  isInsideUnneededParameterizable(unboundGeneric)
+}
+
+private predicate isInsideUnneededCallable(Callable callable) {
+  callable instanceof ConstructedGeneric or
+  isInsideUnneededType(callable.getDeclaringType())
+}
+
+private predicate isInsideUnneededType(Type type) {
+  type instanceof ConstructedType or
+  type.getDeclaringType*() instanceof ConstructedType or
+  type instanceof AnonymousClass or
+  type.getDeclaringType*() instanceof AnonymousClass
+}
+
+private predicate isInsideUnneededParameterizable(Parameterizable parameterizable) {
+  isInsideUnneededCallable(parameterizable) or
+  isInsideUnneededType(parameterizable.(Indexer).getDeclaringType()) or
+  isInsideUnneededType(parameterizable.(DelegateType))
+}
+
+private predicate isCompilerGeneratedParameterizable(Parameterizable parameterizable) {
+  parameterizable.isCompilerGenerated() or
+  parameterizable.getDeclaringType*().isCompilerGenerated()
+}
+
+private predicate isCompilerGeneratedAttributable(Attributable attributable) {
+  // Default parameter values on delegates are in the tree multiple times
+  // due to the compiler generated `Invoke`.
+  isCompilerGeneratedParameterizable(attributable.(Parameter).getDeclaringElement())
+}
```

Comment:
I wonder if all of this could be simplified to:
```ql
predicate isNotNeeded(Element e) {
  e.(Expr).isImplicit()
  or
  e.(Declaration).isCompilerGenerated()
  or
  e instanceof ConstructedGeneric
  or
  e instanceof AnonymousClass
  or
  isNotNeeded(e.getParent())
  or
  isNotNeeded(e.(Attribute).getTarget())
}
```

Reply:
Performance can also be improved by restricting to elements `e` for which `shouldPrint(e)` holds.

Reply:
It would be great if we could convert those predicates to a single more elegant, more readable one. I applied this change, but it results in way too many changes in the expected files. I'll investigate the changes in smaller batches.

One initial finding:
Adding the `.isImplicit()` removes too many things from the trees:
* implicit `this` access, which is okay,
* implicit casts, which is definitely not great, because `10` would also be removed from `byte b = 10;`
* `OperatorCall` to implicit conversion

and maybe others too. So I'll omit this change, and instead I'll probably only handle implicit `this` access. What do you think?

Reply:
Thanks a lot for investigating this; I agree that `isImplicit()` cannot be used then.

Reply:
I pushed a fix for this. The following seems to cover the same cases as the initial PR:
```
private predicate isFilteredCompilerGenerated(Declaration d) {
  d.isCompilerGenerated() and
  not d instanceof Accessor
}

private predicate isNotNeeded(Element e) {
  isFilteredCompilerGenerated(e)
  or
  e instanceof ConstructedGeneric
  or
  e instanceof AnonymousClass
  or
  isNotNeeded(e.(Declaration).getDeclaringType())
  or
  isNotNeeded(e.(Parameter).getDeclaringElement())
  or
  isNotNeeded(e.(Attribute).getTarget())
}
```

Reply:
Excellent @tamasvajk !